File: cli/Sources/TuistXCResultService/TestCaseFailure.swift
import Foundation
import Path

public struct TestCaseFailure {
    public enum IssueType: String {
        case errorThrown = "Thrown Error"
        case assertionFailure = "Assertion Failure"
        case issueRecorded = "Issue Recorded"
    }

    public let message: String?
    public let path: RelativePath?
    public let lineNumber: Int
    public let issueType: IssueType?

    public init(
        message: String?,
        path: RelativePath?,
        lineNumber: Int,
        issueType: IssueType?
    ) {
        self.message = message
        self.path = path
        self.lineNumber = lineNumber
        self.issueType = issueType
    }
}

File: server/lib/tuist_web/api/schemas/tests/test_case_failure.ex
defmodule TuistWeb.API.Schemas.Tests.TestCaseFailure do
  @moduledoc """
  Shared schema for test case failures, used by both test cases and argument variants.
  """
  alias OpenApiSpex.Schema

  require OpenApiSpex

  OpenApiSpex.schema(%{
    title: "TestCaseFailure",
    type: :object,
    description: "A test case failure.",
    properties: %{
      message: %Schema{
        type: :string,
        description: "The failure message."
      },
      path: %Schema{
        type: :string,
        description: "The file path where the failure occurred, relative to the project root."
      },
      line_number: %Schema{
        type: :integer,
        description: "The line number where the failure occurred."
      },
      issue_type: %Schema{
        type: :string,
        description: "The type of issue that occurred.",
        enum: ["error_thrown", "assertion_failure", "issue_recorded"]
      }
    },
    required: [:line_number]
  })
end
